src/lib/arch/aarch64/ExceptionHandler.cc

```diff
@@ -13,462 +10,98 @@ namespace simeng {
 namespace arch {
 namespace aarch64 {
 
-ExceptionHandler::ExceptionHandler(
-    const std::shared_ptr<simeng::Instruction>& instruction, const Core& core,
-    MemoryInterface& memory, std::shared_ptr<OS::SyscallHandler> sysHandler)
-    : instruction_(*static_cast<Instruction*>(instruction.get())),
-      core(core),
-      memory_(memory),
-      sysHandler_(sysHandler) {
-  resumeHandling_ = [this]() { return init(); };
+ExceptionHandler::ExceptionHandler(const Core& core) : core_(core) {}
+
+bool ExceptionHandler::tick() {
+  // If the syscall handler has been envoked, conclude the syscall only once the
+  // result has been returned
+  if (envokingSycallHandler_) {
+    if (!syscallReturned_) return false;
+    return concludeSyscall();
+  }
+
+  // If an instruction ahs been resgitered with an exception, begin its
+  // processing
+  if (instruction_ != nullptr) return handleException();
+
+  // Return false as a default case
+  return false;
+}
+
+void ExceptionHandler::registerException(
+    std::shared_ptr<simeng::Instruction> instruction) {
+  // TODO: look into dynamic_pointer_cast for safer casting
+  instruction_ = std::static_pointer_cast<Instruction>(instruction);
 }
 
-bool ExceptionHandler::tick() { return resumeHandling_(); }
+bool ExceptionHandler::handleException() {
+  if (instruction_ == nullptr) return false;
```

Comment:
is this check needed? given in `tick()` we have the check `if (instruction_ != nullptr) { handleException() }`. I also verified this in all core types where Core::processException always calls `exceptionHandler_.tick()`;

Reply:
Very good spot, removing redundant check